File: 09/python/main.py
def solution(lines):
    lines = [[int(y) for y in x.strip()] for x in lines]
    total = 0
    low_points = []
    for i in range(0, len(lines)):
        for j in range(0, len(lines[i])):
            val = (lines[i][j])
            up = (lines[i-1][j]) if i > 0 else 10
            down = (lines[i+1][j]) if i < len(lines) -1 else 10
            left = (lines[i][j-1]) if j > 0 else 10
            right = (lines[i][j + 1]) if j < len(lines[i]) - 1 else 10
            if val < up and val < down and val < left and val < right:
                total += val + 1
                low_points.append((i, j))
    print(total) # part 1
    
    sizes = []
    for low_point in low_points:
        points_to_check = [low_point]
        points_checked = []


        while len(points_to_check):
            (i,j) = points_to_check.pop()

            val = lines[i][j]
            up = (lines[i-1][j]) if i > 0 else 10
            down = (lines[i+1][j]) if i < len(lines) -1 else 10
            left = (lines[i][j-1]) if j > 0 else 10
            right = (lines[i][j + 1]) if j < len(lines[i]) - 1 else 10

            if up > val and up < 9:
                points_to_check.append((i-1, j))
            
            if down > val and down < 9:
                points_to_check.append((i+1, j))
            
            if left > val and left < 9:
                points_to_check.append((i, j-1))
            
            if right > val and right < 9:
                points_to_check.append((i, j+1))
            
            points_checked.append((i,j))
            points_to_check = [x for x in points_to_check if not x in points_checked]
        
        sizes.append(len(points_checked))
    [a,b,c] =sorted(sizes)[-3:]
    print(a * b * c) # part 2

with open('input.txt') as f:
    lines = f.readlines()
    print(solution(lines))

File: 09/haskell/main.hs
import Data.Array
import Data.Char
import Data.List
import Data.Maybe
import System.IO

type Size = (Int, Int)

type Position = (Int, Int)

-- Uses haskell arrays because we're performing a bunch of index operations
type Board = Array (Int, Int) Int

buildArray :: String -> Board
buildArray value =
  array
    ((1, 1), (w, h))
    [ ((x, y), v)
      | (y, vs) <- zip [1 ..] vss,
        (x, v) <- zip [1 ..] vs
    ]
  where
    h = length vss
    w = length $ head vss
    vss = map (map digitToInt) $ lines value

isLowPoint :: Board -> Position -> Bool
isLowPoint board (x, y) = all ((> value) . (board !)) neighbors
  where
    value = board ! (x, y)
    neighbors = getNeighbors (bounds board) (x, y)

getNeighbors :: ((Int, Int), (Int, Int)) -> Position -> [Position]
getNeighbors ((minX, minY), (maxX, maxY)) (x, y) = catMaybes [left, right, up, down]
  where
    left = if x == minX then Nothing else Just (x -1, y)
    right = if x == maxX then Nothing else Just (x + 1, y)
    up = if y == minY then Nothing else Just (x, y - 1)
    down = if y == maxY then Nothing else Just (x, y + 1)

getBasin :: Board -> Position -> [Position]
getBasin board position =
  if board ! position == 9
    then []
    else position : concatMap (getBasin board) heigherNeighbors
  where
    value = board ! position
    neighbors = getNeighbors (bounds board) position
    heigherNeighbors = filter ((> value) . (board !)) neighbors

getBasinSize :: Board -> Position -> Int
getBasinSize board position = length $ nub $ getBasin board position

solution1 :: String -> Int
solution1 value = foldl foldFunc 0 $ indices board
  where
    board = buildArray value
    foldFunc acc x = if isLowPoint board x then board ! x + acc + 1 else acc

solution2 :: String -> Int
solution2 value = product $ take 3 $ reverse $ sort $ map (getBasinSize board) lowPoints
  where
    board = buildArray value
    lowPoints = filter (isLowPoint board) (indices board)

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  print (solution1 contents)
  print (solution2 contents)
  hClose handle
